pkg/kv/kvserver/kvstorage/init.go

```diff
@@ -291,103 +293,197 @@ func IterateRangeDescriptorsFromDisk(
+	sl := make([]Replica, 0, len(m))
+	for _, repl := range m {
+		sl = append(sl, repl)
 	}
+	sort.Slice(sl, func(i, j int) bool {
+		return sl[i].RangeID < sl[j].RangeID
+	})
+	return sl
+}
+
+func (m ReplicaMap) getOrMake(rangeID roachpb.RangeID) Replica {
+	ent := m[rangeID]
+	ent.RangeID = rangeID
+	return ent
+}
+
+func (m ReplicaMap) setReplicaID(rangeID roachpb.RangeID, replicaID roachpb.ReplicaID) Replica {
+	ent := m.getOrMake(rangeID)
+	ent.ReplicaID = replicaID
+	m[rangeID] = ent
+	return ent
+}
 
-	// TODO(sep-raft-log): if there is any other data that we mandate is present here
-	// (like a HardState), validate that here.
+func (m ReplicaMap) setHardState(rangeID roachpb.RangeID, hs raftpb.HardState) {
+	ent := m.getOrMake(rangeID)
+	ent.hardState = hs
+	m[rangeID] = ent
+}
 
-	return m, nil
+func (m ReplicaMap) setDesc(rangeID roachpb.RangeID, desc roachpb.RangeDescriptor) error {
+	ent := m.getOrMake(rangeID)
+	if ent.Desc != nil {
+		return errors.AssertionFailedf("overlapping descriptors %v and %v", ent.Desc, &desc)
+	}
+	ent.Desc = &desc
+	m[rangeID] = ent
+	return nil
 }
 
 // LoadAndReconcileReplicas loads the Replicas present on this
 // store. It reconciles inconsistent state and runs validation checks.
 //
-// TOOD(sep-raft-log): consider a callback-visitor pattern here.
-func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (*EngineReplicas, error) {
+// TODO(sep-raft-log): consider a callback-visitor pattern here.
+func LoadAndReconcileReplicas(ctx context.Context, eng storage.Engine) (ReplicaMap, error) {
 	ident, err := ReadStoreIdent(ctx, eng)
 	if err != nil {
 		return nil, err
 	}
 
-	initM := map[storage.FullReplicaID]*roachpb.RangeDescriptor{}
+	s := ReplicaMap{}
+
 	// INVARIANT: the latest visible committed version of the RangeDescriptor
 	// (which is what IterateRangeDescriptorsFromDisk returns) is the one reflecting
 	// the state of the Replica.
 	// INVARIANT: the descriptor for range [a,z) is located at RangeDescriptorKey(a).
 	// This is checked in IterateRangeDescriptorsFromDisk.
-	if err := IterateRangeDescriptorsFromDisk(
-		ctx, eng, func(desc roachpb.RangeDescriptor) error {
-			// INVARIANT: a Replica's RangeDescriptor always contains the local Store,
-			// i.e. a Store is a member of all of its local Replicas.
-			repDesc, found := desc.GetReplicaDescriptor(ident.StoreID)
-			if !found {
-				return errors.AssertionFailedf(
-					"RangeDescriptor does not contain local s%d: %s",
-					ident.StoreID, desc)
-			}
+	{
+		var lastDesc roachpb.RangeDescriptor
+		if err := IterateRangeDescriptorsFromDisk(
+			ctx, eng, func(desc roachpb.RangeDescriptor) error {
+				if lastDesc.RangeID != 0 && desc.StartKey.Less(lastDesc.EndKey) {
+					return errors.AssertionFailedf("overlapping descriptors %s and %s", lastDesc, desc)
+				}
+				lastDesc = desc
+
+				if err := s.setDesc(desc.RangeID, desc); err != nil {
+					return err
+				}
+				return nil
+			}); err != nil {
```

Comment:
supernit: I'd put comma here and `); err != nil` check on the next line to avoid the weird indentation.